examples/async_replication/async_replicated_container.h

```diff
@@ -0,0 +1,421 @@
+      reader_port = config.port;
+      delay = config.delay;
+      nodes = config.nodes_list;
+      is_verbose = config.is_verbose;
+      show_network_err = config.show_network_errors;
+      max_waits = config.max_waits;
+  }
+  ~AsyncReplicatedContainer(){
+    stop();
+  }
+
+  void reader(uint16_t port){
+    int waits = 0;
+    bool is_stop = false;
+    Relay buffer;
+
+    while (true) {
+      is_stop = state.ImmutableUse(
+          [](SharedState const&state){ return state.die;});
+      if(is_stop){
+        break;
+      }
+      try {
+        current::net::Socket socket((current::net::BarePort(port)));
+        // FIXME? how to accept multiple connections on same port with SO_REUSEADDR/REUSEPORT
+        current::net::Connection connection(socket.Accept());
+
+        if(is_verbose) {
+          state.MutableUse([port](SharedState& state) {
+            std::cout << "Reader connected on port " << port << std::endl;
+          });
+        }
+
+        while (waits < max_waits) {
+          is_stop = state.ImmutableUse(
+              [](SharedState const&state){ return state.die;});
+          if(is_stop){
+            break;
+          }
+          try {
+            // Get the data size
+            buffer = recv_relay(connection);
+
+            state.MutableUse([&buffer, this](SharedState& state) {
+              std::pair<std::string, int> data(buffer.key, buffer.value);
+              bool is_insert = state.data.find(buffer.key) == state.data.end();
+              bool is_valid_update = (!is_insert) && (buffer.clock > state.clock[buffer.key]);
+
+              // new row
+              if(is_insert || is_valid_update) {
+                state.clock[buffer.key] = buffer.clock;
+                state.data[buffer.key] = buffer.value;
+              }
+              if(is_verbose){
+                if(is_insert){
+                  std::cout << "NEW [" << buffer.replica_id << "] key "<< buffer.key << std::endl;
+                } else if (is_valid_update){
+                  std::cout << "REPLICATED [" << buffer.replica_id << "] key " << buffer.key << std::endl;
+                } else {
+                  std::cout << "IGNORED [" << buffer.replica_id << "] key " << buffer.key << std::endl;
+                }
+              }
+
+            });
+          } catch (const current::Exception &e) {
+            waits += 1;
+            std::this_thread::sleep_for(std::chrono::milliseconds(delay));
+          }
+        }
+        waits = 0;
+      } catch (const current::Exception &e) {
+        if(show_network_err) {
+          state.MutableUse([&e](SharedState& state) {
+            std::cout << "error reader" << ": " << e.OriginalDescription() << std::endl;
+          });
+        }
+      }
+      std::this_thread::sleep_for(std::chrono::milliseconds(delay));
+    }
+    // stopped
+  }
+
+  Relay recv_relay(current::net::Connection &c){
+    // Get buffer length
+    char* len_buf = (char*)malloc(sizeof(size_t));
```

Comment:
A `vector<uint8_t>` (or of any other type, really) is far safer than `malloc`. Although if you just need a `size_t`, declare `size_t len` and `BlockingRead(&len, sizeof(len)` into it.

Reply:
it works for vector<uint8_t> but returns a compilation error for reading directly to int:
```c++
size_t len_buf;
auto size = c->BlockingRead(&len_buf, sizeof(len_buf));
if (!size) throw;
len_buf = be64toh(len_buf);
```
returns
```
error: no matching function for call to ‘current::net::Connection::BlockingRead(size_t*, long unsigned int)’
  178 |     auto size = c->BlockingRead(&len_buf, sizeof(len_buf));
```

Reply:
Currently replaced all `malloc()` calls with vector<uint8_t>